07-routing-gossip.md

```diff
@@ -510,16 +521,196 @@ An endpoint node:
 
 ### Rationale
 
-Once the announcement has been processed, it's added to a list of outgoing
-announcements, destined for the processing node's peers, replacing any older
-updates from the origin node. This list of announcements will be flushed at
+Once the gossip message has been processed, it's added to a list of outgoing
+messages, destined for the processing node's peers, replacing any older
+updates from the origin node. This list of gossip messages will be flushed at
 regular intervals: such a store-and-delayed-forward broadcast is called a
-_staggered broadcast_. Also, such batching of announcements forms a natural rate
+_staggered broadcast_. Also, such batching forms a natural rate
 limit with low overhead.
 
-The sending of all announcements on reconnection is naive, but simple,
+The sending of all gossip on reconnection is naive, but simple,
 and allows bootstrapping for new nodes as well as updating for nodes that
-have been offline for some time.
+have been offline for some time.  The `gossip_queries` option
+allows for more refined synchronization.
+
+## Query Messages
+
+Negotiating the `gossip_queries` option enables a number of extended
+queries for gossip synchronization.  These explicitly request what
+gossip should be received.
+
+There are several messages which contain a long array of
+`short_channel_id`s (called `encoded_short_ids`) so we utilize a
+simple compression scheme: the first byte indicates the encoding, the
+rest contains the data.
+
+Encoding types:
+* `0`: uncompressed array of `short_channel_id` types, in ascending order.
+* `1`: array of `short_channel_id` types, in ascending order, compressed with zlib<sup>[1](#reference-1)</sup>
```

Comment:
Each new encoding type needs a distinct feature bit. Otherwise they're _all_ mandatory. Some clients may not want to package zlib, nor have to implement defensive decoding measures to ensure an enormous amount of memory isn't allocated due to a maliciously crafted payload.

Reply:
Yes, `zlib` and `uncompressed` are mandatory and we'll probably never need to add another format (`zlib` already does a fairly good job).

Reply:
I mean why should zlib itself be mandatory? IMO it was a bit of a premature optimization (a few additional round trips really aren't that bad). It _forces_ implementations to implement defensive parsing strategies, or it may become a DoS vector if the spec doesn't explicitly pay out the risks and mitigation strategies.

Reply:
zlib is pretty established and robust, and it's now deployed and thus too late :(

Any other compression methods would require a new feature bit.  For now, don't advertise gossip_queries if you don't support receiving zlib.

@Roasbeef pointed out that feature bits also allow us to do fine-grained disable if problems are found with an implementation.  **Let this be a lesson-learned**: every feature needs a feature bit, even if everyone intends to implement it!

Reply:
Meanwhile, I've added a paragraph on worst-case behavior:

```
Note that a 65535-byte zlib message can decompress into 67632120
bytes<sup>[2](#reference-2)</sup>, but since the only valid contents 
are unique 8-byte values, no more than 14 bytes can be duplicated
across the stream: as each duplicate takes at least 2 bits, no valid
contents could decompress to more then 3669960 bytes.
```

This number is very conservative: you still need some literals, and you still need the two-bytes between the 14 duplicated bytes, etc.